### ISSUE TITLE
rtx -ls hangs forever

### ISSUE THREAD

**Issue by amoosbr** [jdxcode/rtx#547]
**Describe the bug**
When I add `azure-functions-core-tools = "4"` to `.rtx.toml`, I can no longer successfully run `rtx ls -current`.
It always hangs indefinitely (Update: Actually, waited now longer and it just got killed after 8 minutes with exit code 137).
Neither the tool nor the plugin for `azure-functions-core-tools` needs to be installed. Also tried with the plugin and tool installed, but this didn't change anything.

I really have no idea why and if this is now related to my local environment, as I tried various things to understand this and at some point in time using my own fork of the `azure-functions-core-tools` asdf plugin.

I don't have this error, when just i.e. `shfmt = "3"` is in `.rtx.toml`, although I don't have the plugin and tool installed as well.
I only saw it with `azure-functions-core-tools = "4"` yet.

**To Reproduce**

.rtx.toml
```toml
[tools]
azure-functions-core-tools = "4"
shfmt = "3"
```

```sh
❯ cd ~/dev/tmp/rtx-azure-functions/not-working
[WARN] Tools not installed: azure-functions-core-tools@4, shfmt@3
❯ RTX_TRACE=1 rtx ls --current
[TRACE] (1) rtx::config::config_file::rtx_toml: [src/config/config_file/rtx_toml.rs:59] parsing: /Users/user/.config/rtx/config.toml
[TRACE] (1) rtx::plugins::rtx_plugin_toml: [src/plugins/rtx_plugin_toml.rs:29] parsing: /Users/user/.local/share/rtx/plugins/java/rtx.plugin.toml
[TRACE] (1) rtx::plugins::rtx_plugin_toml: [src/plugins/rtx_plugin_toml.rs:29] parsing: /Users/user/.local/share/rtx/plugins/ruby/rtx.plugin.toml
[TRACE] (3) rtx::config::config_file::rtx_toml: [src/config/config_file/rtx_toml.rs:59] parsing: /Users/user/.config/rtx/config.toml
[TRACE] (11) rtx::config::config_file::rtx_toml: [src/config/config_file/rtx_toml.rs:59] parsing: /Users/user/Projects/tmp/rtx-azure-functions/not-working/.rtx.toml
[TRACE] (1) rtx::config: [src/config/mod.rs:69] Settings: Settings {
    experimental: true,
    missing_runtime_behavior: Prompt,
    always_keep_download: false,
    always_keep_install: false,
    legacy_version_file: true,
    plugin_autoupdate_last_check_duration: 604800s,
    verbose: false,
    asdf_compat: false,
    jobs: 4,
    shorthands_file: None,
    disable_default_shorthands: false,
    log_level: Trace,
    raw: false,
}
[TRACE] (11) rtx::cache: [src/cache.rs:78] reading ~/Library/Caches/rtx/golang/legacy_filenames.msgpack.z
[DEBUG] Files: ~/Projects/tmp/rtx-azure-functions/not-working/.rtx.toml, ~/.config/rtx/config.toml
Installed Plugins: golang, java, ruby
[DEBUG] rtx ls --current
[TRACE] (11) rtx::cache: [src/cache.rs:78] reading ~/Library/Caches/rtx/java/aliases.msgpack.z
[TRACE] (5) rtx::cache: [src/cache.rs:78] reading ~/Library/Caches/rtx/golang/remote_versions.msgpack.z
[TRACE] (12) rtx::cache: [src/cache.rs:78] reading ~/Library/Caches/rtx/ruby/remote_versions.msgpack.z
[TRACE] (8) rtx::cache: [src/cache.rs:78] reading ~/Library/Caches/rtx/node/remote_versions.msgpack.z
[TRACE] (4) rtx::cache: [src/cache.rs:78] reading ~/Library/Caches/rtx/python/remote_versions.msgpack.z
[TRACE] (11) rtx::cache: [src/cache.rs:78] reading ~/Library/Caches/rtx/java/remote_versions.msgpack.z
[DEBUG] azure-functions-core-tools@4, shfmt@3, node@lts node@latest, python@latest!-0.1 python@latest, java@temurin-17, golang@prefix:1.20, ruby@3.2

HANGS here 8 minutes

❯ echo $?
137

```

**Expected behavior**

`rtx ls` or `rtx ls --current` working as expected or returning with a error message. Ideally, the error message would allow a developer to fix the plugin and to get `rtx ls` to work again.

**`rtx doctor` output**

```
❯ rtx doctor
rtx version:
  1.29.7 macos-arm64 (a540ac6 2023-05-10)

build:
  Target: aarch64-apple-darwin
  Features: BREW
  Built: Wed, 10 May 2023 16:25:07 +0000
  Rust Version: rustc 1.69.0 (84c898d65 2023-04-16)
  Profile: release

shell:
  /bin/zsh
  zsh 5.9 (x86_64-apple-darwin22.0)

rtx environment variables:
  RTX_USE_TOML=1
  RTX_SHELL=zsh

settings:
  {"experimental": "true", "missing_runtime_behavior": "prompt", "always_keep_download": "false", "always_keep_install": "false", "legacy_version_file": "true", "plugin_autoupdate_last_check_duration": "10080", "trusted_config_paths": "[\"~/sth\"]", "verbose": "false", "asdf_compat": "false", "jobs": "4", "disable_default_shorthands": "false", "log_level": "INFO", "raw": "false"}

config files:
  /Users/user/.config/rtx/config.toml
  /Users/user/dev/tmp/rtx-azure-functions/not-working/.rtx.toml

plugins:
  golang   https://github.com/kennyp/asdf-golang.git#f006a12
  java     https://github.com/rtx-plugins/rtx-java.git#24706b7
  node     (core)
  python   (core)
  ruby     https://github.com/rtx-plugins/rtx-ruby.git#e0728ce

toolset:
  azure-functions-core-tools@4, shfmt@3, node@lts node@latest, python@latest!-0.1 python@latest, java@temurin-17, golang@prefix:1.20, ruby@3.2

2 problems found:
plugin azure-functions-core-tools is not installed

plugin shfmt is not installed
```

**Comment by jdxcode** [comment]
I actually got a panic when I used this plugin, does `rtx ls --json` work?

**Comment by jdxcode** [comment]
this may be related to #539, I suspect github may have been timing out since it's been having a lot of issues lately

**Comment by amoosbr** [comment]
> I actually got a panic when I used this plugin, does `rtx ls --json` work?

Yes, `rtx ls --json` is working:
```sh
❯ rtx ls --json
{
  "azure-functions-core-tools": [
    {
      "version": "4",
      "requested_version": "4",
      "install_path": "/Users/user/.local/share/rtx/installs/azure-functions-core-tools/4",
      "source": {
        "type": ".rtx.toml",
        "path": "/Users/user/dev/tmp/rtx-azure-functions/not-working/.rtx.toml"
      }
    }
  ],
  ...
}
```

**Comment by amoosbr** [comment]
> this may be related to #539, I suspect github may have been timing out since it's been having a lot of issues lately

Of course, I don't know exactly, what rtx is doing for an `rtx ls`, but I thought it is not related to network issues.
Especially, as I don't have the `azure-functions-core-tools` plugin installed at all.
I thought it would just list a missing tool version. Similar to `shfmt` in this example. For `shfmt`, I also have no plugin installed.

```sh
❯ rtx ls --current
shfmt  3 (missing)     ~/dev/tmp/rtx-azure-functions/working/.rtx.toml 3
```

In addition, when I install the plugin, run `list-all` of the plugin it is quite fast. But a second later `rtx ls --current` hangs again. But perhaps rtx is calling sth else in addition to `list-all`. This is something, I wasn't able to figure out yet.

```sh
❯ cd ../not-working
[WARN] Tools not installed: azure-functions-core-tools@4, shfmt@3
❯ rtx plugins install azure-functions-core-tools
rtx azure-functions-core-tools https://github.com/daveneeley/asdf-azure-functions-core-tools.git#✓ 1s %
❯ time ~/.local/share/rtx/plugins/azure-functions-core-tools/bin/list-all
0.10 0.91 0.92 0.93 0.94 0.95 0.96 0.97 0.98 0.99 0.100 1.0 1.0.1 1.0.3 1.0.4 1.0.6 1.0.7 1.0.8 1.0.9 1.0.10 1.0.11 1.0.12 1.0.14 1.0.15 1.0.17 1.0.18 1.0.19 1.0.20 1.0.25 1.0.27 1.0.28 2.0.1-beta.8 2.0.1-beta.22 2.0.1-beta.23 2.0.1-beta.24 2.0.1-beta.25 2.0.1-beta.26 2.0.1-beta.27 2.0.1-beta.28 2.0.1-beta.29 2.0.1-beta.30 2.0.1-beta.31 2.0.1-beta.32 2.0.1-beta.33 2.0.1-beta.34 2.0.1-beta.35 2.0.1-beta.36 2.0.1-beta.37 2.0.1-beta.38 2.0.1-beta.39 2.0.2 2.0.3 2.0.1948 2.0.2359 2.1.725 2.1.746 2.1.748 2.2.32 2.2.70 2.3.148 2.3.199 2.4.249 2.4.290 2.4.299 2.4.317 2.4.401 2.4.419 2.4.498 2.4.499 2.5.553 2.6.666 2.6.1048 2.6.1064 2.6.1071 2.6.1080 2.6.1106 2.7.1149 2.7.1158 2.7.1373 2.7.1415 2.7.1468 2.7.1471 2.7.1480 2.7.1505 2.7.1524 2.7.1558 2.7.1575 2.7.1585 2.7.1633 2.7.1704 2.7.1724 2.7.1812 2.7.1846 2.7.1948 2.7.2045 2.7.2100 2.7.2184 2.7.2254 2.7.2359 2.7.2508 2.7.2628 2.7.2748 2.7.2796 2.7.2855 2.7.2883 2.7.2913 2.7.2936 2.7.3023 2.7.3188 3.0.1740 3.0.1975 3.0.2009 3.0.2106 3.0.2245 3.0.2358 3.0.2534 3.0.2630 3.0.2750 3.0.2798 3.0.2852 3.0.2881 3.0.2912 3.0.2931 3.0.2996 3.0.3160 3.0.3216 3.0.3233 3.0.3245 3.0.3284 3.0.3331 3.0.3354 3.0.3388 3.0.3442 3.0.3477 3.0.3528 3.0.3568 3.0.3734 3.0.3785 3.0.3873 3.0.3904 3.0.4378 3.0.4425 3.0.4484 3.0.4502 3.0.4585 3.0.4626 3.0.4669 3.0.4727 3.0.4753 3.0.4806 3.0.4837 3.0.4868 3.0.4899 4.0.3780 4.0.3893 4.0.3928 4.0.3971 4.0.4385 4.0.4426 4.0.4483 4.0.4544 4.0.4590 4.0.4629 4.0.4653 4.0.4670 4.0.4704 4.0.4736 4.0.4785 4.0.4829 4.0.4865 4.0.4895 4.0.4915 4.0.5030 4.0.5085 4.0.5095 4.0.5110 4.0.5148
~/.local/share/rtx/plugins/azure-functions-core-tools/bin/list-all  0.02s user 0.03s system 11% cpu 0.471 total
❯ rtx ls --current

# hangs here
```

**Comment by amoosbr** [comment]
Not sure, if it is relevant, but calling `rtx current` is working as expected:

```sh
❯ rtx current
[WARN] azure-functions-core-tools@4 is specified in ~/Projects/tmp/rtx-azure-functions/not-working/.rtx.toml, but not installed
azure-functions-core-tools 4
[WARN] shfmt@3 is specified in ~/Projects/tmp/rtx-azure-functions/not-working/.rtx.toml, but not installed
shfmt 3
node 18.16.0 20.1.0
```

**Comment by amoosbr** [comment]
OK, a workaround for this issue is to rename the plugin in the `.rtx.toml` file.

Using the following config, `rtx ls` and `rtx ls --current` immediately start working again (linking to the same azure-functions repository the `azure-functions-core-tools` should be using.
So I think the issue is not really plugin code itself, but perhaps related to the shorthand handling or sth like this in rtx?
Not sure.

Non-blocking `.rtx.toml` config:
```toml
[tools]
azure-functions = "4"
shfmt = "3"

[plugins]
azure-functions = 'https://github.com/daveneeley/asdf-azure-functions-core-tools.git'
```

**Comment by amoosbr** [comment]
This issue resolved itself with latest 1.30.x release. I expect #555 resolved this. Thanks.

**Comment by jdxcode** [comment]
I kind of doubt #555 had an impact, though I also don't understand why `--current` would have an affect. This is all it does and that bit of code should not be capable of causing a hang: https://github.com/jdxcode/rtx/blob/c6680500e0e410e577510d375c98ab80ce521a03/src/cli/ls.rs#L58

I suppose we can just wait and see if it crops up again.
